examples/matching_engine/src/lib.rs

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use async_graphql::{scalar, InputObject, Request, Response, SimpleObject};
+use fungible::{AccountOwner, FungibleTokenAbi};
+use linera_sdk::base::{Amount, ApplicationId, ContractAbi, ServiceAbi};
+use linera_views::{common::CustomSerialize, views::ViewError};
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeSet;
+
+// TODO(#768): Remove the derive macros.
+#[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Serialize, Deserialize)]
+pub struct MatchingEngineAbi;
+
+impl ContractAbi for MatchingEngineAbi {
+    type InitializationArgument = ();
+    type Parameters = Parameters;
+    type Operation = Operation;
+    type ApplicationCall = ApplicationCall;
+    type Message = Message;
+    type SessionCall = ();
+    type Response = ();
+    type SessionState = ();
+}
+
+impl ServiceAbi for MatchingEngineAbi {
+    type Parameters = Parameters;
+    type Query = Request;
+    type QueryResponse = Response;
+}
+
+// Buying or selling items 1 with 0.
+// Forgetting about types and units, if we do a transaction (say a buy) we will have
+// Account[0] -= price * quantity
+// Account[1] += quantity
+// Thus the quantity (also named Count) has to be an Amount.
+//
+// When we has ask > bid then the winner for the residual cash is the liquidity
+// provider.
+#[derive(
+    Clone, Copy, Debug, PartialEq, PartialOrd, Deserialize, Serialize, SimpleObject, InputObject,
+)]
+#[graphql(input_name = "PriceInput")]
+pub struct Price {
+    pub price: u64,
+}
+
+impl CustomSerialize for Price {
+    fn to_custom_bytes(&self) -> Result<Vec<u8>, ViewError> {
+        let mut short_key = bcs::to_bytes(&self.price)?;
+        short_key.reverse();
+        Ok(short_key)
+    }
+
+    fn from_custom_bytes(short_key: &[u8]) -> Result<Self, ViewError> {
+        let mut bytes = short_key.to_vec();
+        bytes.reverse();
+        let value = bcs::from_bytes(&bytes)?;
+        Ok(value)
+    }
+}
+
+impl Price {
+    pub fn revert(&self) -> Self {
+        Price {
+            price: u64::MAX - self.price,
+        }
+    }
+}
+
+pub fn product_price_amount(price: Price, count: Amount) -> Amount {
+    let mut result = Amount::zero();
+    for _ in 0..price.price {
```

Comment:
We could keep it a `u64`, just change the meaning of it to make it a fixed-point fraction:

Instead of multiplying the count with `price`, we multiply with `price` and then divide by 10000. that means the price is in 1/10000ths of token 0, instead of whole tokens 0.

Reply:
We could do that, but that breaks the non-divisibility of the tokens.
So, when we do the transaction, we have actually to do an Euclidean division.
Is this really what we want?

Reply:
That's true, it makes it a bit more complicated.
Maybe we should only allow orders where the token1 count is divisible by 10000. That would solve the problem.

But I'm also okay with the current approach. We should just clearly document then that it only makes sense if token1 is a few orders of magnitude more valuable than token0.

Reply:
Ok, I added documentation on this aspect of the problem.